### ISSUE TITLE
Message placeholders, context, and dynamic properties

### ISSUE THREAD

**Issue by WillBAnders** [flash-labs/FlashLibs#2]
## Description

Placeholders allow messages to access and include information that is not specifically provided by the developer as an argument to the function. For example, if a command is run by a player placeholders can be used to get information like the player's world or position through properties. These can then be combined as needed to result in a much more power placeholder system than existing methods.

#### Examples

 - `${@player.world}`
    - Retrieves the argument `player` and gets their world (printing the name)
 - `${#player.position.x}`
    - Retrieves the player from the context and gets the x coordinate of their position
 - `${#date}`
    - Retrieves the current date from the context, formatted as `yyyy-mm-dd`

---

### Motivation

Placeholders are a powerful way to customize messages and open up a lot of possibilities for the user. A system like this is necessary to maintain good performance with so many options, and makes it easy for developers to support such behavior. If it were possible to handle placeholders with arguments, then grammatical cases like a/an and singular/plural could be handled as well.

**Comment by hqppyz** [comment]
> If it were possible to handle placeholders with arguments, then grammatical cases like a/an and singular/plural could be handled as well.

Out of curiosity, how would placeholders help grammatical cases such those?

**Comment by WillBAnders** [comment]
Good point, I should clarify things there. The idea was you could do something like `${@item.quantity.singular[cactus, cacti]}` - if the quantity was singular then `cactus` would be used, otherwise `cacti`. Something simple like `a/an` could have a builtin, but for more complex cases this would be necessary.

What's interesting to note is that in some languages like Russian, there are multiple forms of singular and plural (1, 2-4, and others). It's this type of crazy behavior that makes placeholders a difficult problem as the behavior depends on the locale.
